src/VisualStudio/Core/Def/Snippets/AbstractSnippetExpansionClientFactory.cs

```diff
@@ -0,0 +1,24 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Snippets;
+
+internal abstract class AbstractSnippetExpansionClientFactory : ISnippetExpansionClientFactory
+{
+    protected abstract AbstractSnippetExpansionClient CreateSnippetExpansionClient(ITextView textView, ITextBuffer subjectBuffer);
+
+    public AbstractSnippetExpansionClient GetSnippetExpansionClient(ITextView textView, ITextBuffer subjectBuffer)
+    {
+        if (!textView.Properties.TryGetProperty(typeof(AbstractSnippetExpansionClient), out AbstractSnippetExpansionClient expansionClient))
+        {
+            expansionClient = CreateSnippetExpansionClient(textView, subjectBuffer);
+            textView.Properties.AddProperty(typeof(AbstractSnippetExpansionClient), expansionClient);
+        }
```

Comment:
GetOrCreateSingleton doesn't support a non-capturing caller, so I kept the pattern we were already using.

Reply:
➡️ Left this one as-is

Reply:
What's the problem? I don't care if this would allocate a lambda -- the allocation cost here is unmeasurable...

Reply:
I ended up keeping this one not because I think the allocation cost would be a problem, but because the old code used the current pattern and I can't be sure the allocation cost wouldn't be a problem.